🤖 Machine Learning &nbsp;·&nbsp; *Bài 21/23*

[« Machine Learning - Bootstrap Aggregation (Bagging)](020_ml_bootstrap_aggregation.ipynb) &nbsp;•&nbsp; [📚 Mục lục](../README.md) &nbsp;•&nbsp; [Machine Learning - AUC - ROC Curve »](022_ml_auc_roc_curve.ipynb)

# Machine Learning - Cross Validation

> 📖 **Học song song trên W3Schools:** [python_ml_cross_validation.asp](https://www.w3schools.com/python/python_ml_cross_validation.asp)
>
> 🎯 **Trong bài này:** Cross Validation · K-Fold · Stratified K-Fold · Leave-One-Out (LOO) · Leave-P-Out (LPO) · Shuffle Split · Ending Notes

## Cross Validation — Kiểm định chéo

Khi điều chỉnh mô hình, ta muốn **tăng hiệu năng tổng thể trên dữ liệu chưa thấy**. Điều chỉnh tham số có thể cho kết quả tốt hơn nhiều trên tập test — nhưng tối ưu tham số **theo tập test** sẽ làm **rò rỉ thông tin**, khiến mô hình kém hơn trên dữ liệu thực.

Để tránh, nên đánh giá **trước khi** dùng tập test. **Cross validation (CV)** là cách làm điều đó: chia dữ liệu huấn luyện thành nhiều phần, lần lượt dùng từng phần để kiểm tra.

Ta dùng tập dữ liệu **iris**:

In [1]:
from sklearn import datasets

X, y = datasets.load_iris(return_X_y=True)

Có nhiều phương pháp cross validation — bắt đầu với **k-fold**.

## K-Fold

Dữ liệu được chia thành **k** tập (fold) bằng nhau. Mô hình được huấn luyện trên **k-1** fold và kiểm tra trên fold còn lại. Lặp lại **k** lần, mỗi lần một fold làm tập kiểm tra:

In [2]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import KFold, cross_val_score

clf = DecisionTreeClassifier(random_state=42)

k_folds = KFold(n_splits=5)

scores = cross_val_score(clf, X, y, cv=k_folds)

print("Cross Validation Scores: ", scores)
print("Average CV Score: ", scores.mean())
print("Number of CV Scores used in Average: ", len(scores))

Cross Validation Scores:  [1.         1.         0.83333333 0.93333333 0.8       ]
Average CV Score:  0.9133333333333333
Number of CV Scores used in Average:  5


## Stratified K-Fold

Khi các lớp **mất cân bằng**, cần tính đến điều đó khi chia. **Stratified K-Fold** đảm bảo mỗi fold có **tỉ lệ các lớp giống nhau** như toàn bộ dữ liệu:

In [3]:
from sklearn.model_selection import StratifiedKFold

sk_folds = StratifiedKFold(n_splits=5)

scores = cross_val_score(clf, X, y, cv=sk_folds)

print("Cross Validation Scores: ", scores)
print("Average CV Score: ", scores.mean())
print("Number of CV Scores used in Average: ", len(scores))

Cross Validation Scores:  [0.96666667 0.96666667 0.9        0.93333333 1.        ]
Average CV Score:  0.9533333333333334
Number of CV Scores used in Average:  5


## Leave-One-Out (LOO)

Thay vì chọn số fold, **Leave-One-Out** dùng **1** mẫu để kiểm tra và **n-1** mẫu còn lại để huấn luyện — lặp lại cho **từng** mẫu:

In [4]:
from sklearn.model_selection import LeaveOneOut

loo = LeaveOneOut()

scores = cross_val_score(clf, X, y, cv=loo)

print("Average CV Score: ", scores.mean())
print("Number of CV Scores used in Average: ", len(scores))

Average CV Score:  0.94
Number of CV Scores used in Average:  150


> Số điểm số bằng số mẫu trong tập dữ liệu (150).

## Leave-P-Out (LPO)

Giống LOO nhưng chọn **p** mẫu để kiểm tra. Số lần lặp tăng rất nhanh theo p:

In [5]:
from sklearn.model_selection import LeavePOut

lpo = LeavePOut(p=2)

scores = cross_val_score(clf, X, y, cv=lpo)

print("Average CV Score: ", scores.mean())
print("Number of CV Scores used in Average: ", len(scores))

Average CV Score:  0.9382997762863534
Number of CV Scores used in Average:  11175


> 📌 Với p = 2 và 150 mẫu, có tới **11 175** lần huấn luyện!

## Shuffle Split

Khác k-fold, **ShuffleSplit** chừa lại một **tỉ lệ phần trăm** dữ liệu không dùng cho train/test, và lấy mẫu ngẫu nhiên mỗi lần:

In [6]:
from sklearn.model_selection import ShuffleSplit

ss = ShuffleSplit(train_size=0.6, test_size=0.3, n_splits=5, random_state=0)

scores = cross_val_score(clf, X, y, cv=ss)

print("Cross Validation Scores: ", scores)
print("Average CV Score: ", scores.mean())
print("Number of CV Scores used in Average: ", len(scores))

Cross Validation Scores:  [0.97777778 0.91111111 0.91111111 0.95555556 0.95555556]
Average CV Score:  0.9422222222222223
Number of CV Scores used in Average:  5


## Ending Notes — Ghi chú

Đây chỉ là vài phương pháp CV phổ biến. Phương pháp phù hợp phụ thuộc vào **tập dữ liệu** của bạn.

---

## 🧠 Ghi nhớ nhanh

> - Cross validation: chia dữ liệu nhiều lần để đánh giá ổn định, **trước** khi dùng tập test.
> - `cross_val_score(model, X, y, cv=...)`.
> - `KFold` · `StratifiedKFold` (giữ tỉ lệ lớp) · `LeaveOneOut` · `LeavePOut` · `ShuffleSplit`.

---

## 📝 Exercise — Bài tập nhanh

**❓ Câu hỏi:** Phương pháp nào đảm bảo mỗi fold có **tỉ lệ các lớp** giống toàn bộ dữ liệu?

- **A.** KFold
- **B.** LeaveOneOut
- **C.** StratifiedKFold
- **D.** ShuffleSplit

<details>
<summary>👉 Xem đáp án</summary>

✅ **Đáp án: C** — StratifiedKFold

</details>

---

[« Machine Learning - Bootstrap Aggregation (Bagging)](020_ml_bootstrap_aggregation.ipynb) &nbsp;•&nbsp; [📚 Mục lục](../README.md) &nbsp;•&nbsp; [Machine Learning - AUC - ROC Curve »](022_ml_auc_roc_curve.ipynb)